enterprise_gateway/services/processproxies/spark_operator.py

```diff
@@ -22,3 +28,61 @@ def __init__(self, kernel_manager: RemoteKernelManager, proxy_config: dict):
         self.group = "sparkoperator.k8s.io"
         self.version = "v1beta2"
         self.plural = "sparkapplications"
+
+    def get_initial_states(self) -> set:
+        """Return list of states in lowercase indicating container is starting (includes running)."""
+        return ["submitted", "pending", "running"]
+
+    def get_container_status(self, iteration: int | None) -> str:
+        """Determines if container is still active.
+
+        Submitting a new kernel application to the spark operator
+        will take a while to be Running and can also fail before it
+        can actually start any pods. So we will validate that the
+        application succeeds on the Spark Operator side and then
+        delegate to parent to check if the application pod is running.
+
+        Returns
+        -------
+        Empty string if the container cannot be found otherwise.
+        The pod application status in case of success on Spark Operator side
+        Or the retrieved spark operator submission status in other cases (e.g. Failed)
+        """
+        with suppress(Exception):
+            custom_resource = client.CustomObjectsApi().get_namespaced_custom_object(
```

Comment:
How is this guaranteeing that the CRD relative to _this_ kernel is being returned?  Shouldn't this be using a [label selector like the superclass](https://github.com/jupyter-server/enterprise_gateway/blob/185f73bb5701cb0433790cd857a3a56d5da053ec/enterprise_gateway/services/processproxies/k8s.py#L83-L86)?

In fact, what is the purpose of this method?  We should get the status once.  Is the pod status from k8s_process_proxy not sufficient?

Reply:
If I comment out `get_container_status()` I still get properly started operators in the same amount of time with more consistent log statements.

Reply:
```        Submitting a new kernel application to the spark operator
        will take a while to be Running and can also fail before it
        can actually start any pods. So we will validate that the
        application succeeds on the Spark Operator side and then
        delegate to the parent to check if the application pod is running.
```

So, basically, this handles when the crd submission fails, and allow us to fail fast instead of waiting for the timeout period

Reply:
This method isn't failing fast.  Its caller, in the `ContainerProcessProxy` is what is deciding to fail fast.  This is ONLY logging a debug message that isn't providing any value (and skipping the display of useful troubleshooting information that could be provided prior to the CRD reaching the "running" state).

I really don't see what value this is providing, but, if it must be retained, this should be moved to the `CustomResourceProcessProxy` since there's literally nothing specific to Spark Application CRDs.

Reply:
When using CRD, the following happens -> Client --> Submit CRD -> Spark Operator -> Kubernetes -> POD.

The existing code in main today does not identify when a CRD fails (e.g. is rejected for malformation) and never creates a pod. In this case, the code in `ContainerProcessProxy` keeps waiting for the pod to be in a running state for a long period. With these changes, when we identify that the CRD submission failed, we pass to ContainerProcessProxy the failed status and it then fails right away, and when the CRD is processed, we then continue to wait for the pod to be in running state.

As for the location of the code, now I understand you suggestion, I am moving that to CRD.

Reply:
Thanks, this is helpful. I keep forgetting that CRDs are intermediate objects.